Artifacts/Download-Artifacts.ps1

```diff
@@ -33,6 +33,90 @@ function Download-Artifacts {
         [int]    $timeout = 300
     )
 
+    function DownloadPackage {
+        Param(
+            [string] $artifactUrl,
+            [string] $destinationPath,
+            [int]    $timeout = 300
+        )
+
+        $tmpFolder = Join-Path ([System.IO.Path]::GetDirectoryName($destinationPath)) ([System.IO.Path]::GetRandomFileName())
+        $zipFile = Join-Path ([System.IO.Path]::GetTempPath()) "$([Guid]::NewGuid().ToString()).zip"
+        if ($bcContainerHelperConfig.ArtifactsFeedOrganizationAndProject -and ($artifactUrl -like 'https://bcartifacts.*' -or $artifactUrl -like 'https://bcinsider.*')) {
+            # artifactUrl should be either:
+            # https://storageaccount.azureedge.net/type/version/country
+            # or
+            # https://storageaccount.blob.core.windows.net/type/version/country
+            if (-not ($artifactUrl -match '^https:\/\/([^.]+)\..*\.net\/([^\/]+)\/([^\/]+)\/([a-zA-Z0-9]+)$')) {
+                throw "Invalid artifact url '$artifactUrl'"
+            }
+            $organization, $project = $bcContainerHelperConfig.ArtifactsFeedOrganizationAndProject.Split('/')
+            New-Item -Path $tmpFolder -ItemType Directory | Out-Null
+            $artifactVersion = [System.Version]"$($Matches[3])"
+            $feed = $Matches[1]
+            $packageName = "$($Matches[2]).$($Matches[4]).$($artifactVersion.Major)"
+            $packageVersion = "$($artifactVersion.Minor).$($artifactVersion.Build).$($artifactVersion.Revision)"
+            Write-Host "Using az artifacts universal download --organization $organization --project=$project --scope project --feed $feed --name $packageName --version $packageVersion"
+            az artifacts universal download --organization $organization --project=$project --scope project --feed $feed --name $packageName --version $packageVersion --path $tmpFolder
+        }
+        else {
+            $retry = $false
+            do {
+                Download-File -sourceUrl $artifactUrl -destinationFile $zipFile -timeout $timeout
+                Write-Host "Unpacking artifact to tmp folder " -NoNewline
+                try {
+                    Expand-7zipArchive -Path $zipFile -DestinationPath $tmpFolder -use7zipIfAvailable:(!$retry)
+                    $retry = $false
+                }
+                catch {
+                    Remove-Item -path $zipFile -force
+                    if (Test-Path $tmpFolder) {
+                        Remove-Item $tmpFolder -Recurse -Force
```

Comment:
Isn't there a case when `$tmpFolder` has been deleted, but the there is a retry? Will there be side effects from this?

Reply:
This is the "old" code - expand-7zip creates the folder, so it needs to be deleted before retrying